### PR TITLE
revise visited_list_pool.h

### PR DESCRIPTION
**Pull request by intstellar**
some suggestions of revise reset function, hope it is useful.

### CODE REVIEW

File: hnswlib/visited_list_pool.h [hunk 1]
```diff
@@ -13,16 +13,17 @@ namespace hnswlib {
...
             numelements = numelements1;
-            mass = new vl_type[numelements];
+            mass = new vl_type[numelements](); //init = 0
         }
 
         void reset() {
-            curV++;
-            if (curV == 0) {
-                memset(mass, 0, sizeof(vl_type) * numelements);
+            if(curV < 65535){ //avoid overflow of curV and mass (this will cause ambiguity)
```
Comment (yurymalkov) [inline hnswlib/visited_list_pool.h:22]: I think it can go bad if `vl_type` changes (e.g. to a single byte, the number will be 255).

**Review by yurymalkov — CHANGES_REQUESTED** [review]
Thank you! I think there might be problems if the data type is changed.

**Comment by intstellar** [comment]
Of course, you are right. If replace 65535 with std::numeric_limits<unsigned short int>, it should be better.In many case, USHRT_MAX = (short)-1, so your version usually is correct.  I test hnsw in MNIST dataset in my i7-6700 desktop, here are the results, HNSWBuilding index:efConstruction = 48, M = 16Build time: 1.606658stesting...ef=8, recall=0.959300, time=129.919msef=9, recall=0.960800, time=105.947msef=10, recall=0.961500, time=110.346msef=11, recall=0.961700, time=115.865msef=12, recall=0.961800, time=121.701msef=13, recall=0.963000, time=126.229msef=14, recall=0.963500, time=130.206msef=15, recall=0.963600, time=135.056msef=16, recall=0.964100, time=142.711msef=17, recall=0.964300, time=145.094msef=18, recall=0.965000, time=149.804msef=19, recall=0.964900, time=154.744msef=20, recall=0.965200, time=166.172msef=21, recall=0.965800, time=169.351msef=22, recall=0.965900, time=174.322msef=23, recall=0.966400, time=179.656msef=24, recall=0.966400, time=184.361msef=25, recall=0.966700, time=186.775msef=26, recall=0.967000, time=191.210msef=27, recall=0.967200, time=199.543msef=28, recall=0.966900, time=200.513msef=29, recall=0.967000, time=204.763msef=30, recall=0.967200, time=213.208msef=31, recall=0.967300, time=214.961msActual memory usage: 117Mb  KNN with K=39717/10000, time used:20.342s  We can find that HNSW is 12+ times faster than KNN even in a small dataset. Excellent work! And the usage of lock & mutex is impressive, simple but effective. I think this project is still in progress, for example, in addPoint function, "if the element with the same label already exists, it will be updated instead of creating a new element". like this: auto search = label_lookup_.find(label);   if (search != label_lookup_.end()) {   tableint existingInternalId = search->second;    templock_curr.unlock();    std::unique_lock <std::mutex> lock_el_update(link_list_update_locks_[(existingInternalId & (max_update_element_locks - 1))]);   updatePoint(data_point, existingInternalId, 1.0);   return existingInternalId;} This may not exactly what we want. May be it should like this: if (WHEN_WE_WANT && search != label_lookup_.end()) {   ...} Thank you.05.03.2021, 00:44, "Yury Malkov" <notifications@github.com>:
@yurymalkov requested changes on this pull request.

Thank you! I think there might be problems if the data type is changed.

In hnswlib/visited_list_pool.h:
>          }
 
         void reset() {
-            curV++;
-            if (curV == 0) {
-                memset(mass, 0, sizeof(vl_type) * numelements);
+            if(curV < 65535){ //avoid overflow of curV and mass (this will cause ambiguity)

I think it can go bad if vl_type changes (e.g. to a single byte, the number will be 255).

**Comment by yurymalkov** [comment]
Thank you! Looks good to me!
Will merge it after we figure our why CI flag is not showing in github web interface.

**Comment by intstellar** [comment]
Hi Yury,I newly put forward some requests. The follwing is the changelog. If you find some of them are useful, you can merge them anyway you like. If you think them are useless, just discard them. Thank you for your time.  changelog0316 1. add an indicator(bUpdate) at addPoint function, so it will only call updatePoint when we need.2. redefine CompareByFirst, let the smaller at the top, then we needn't use minus distance at std::priority_queue.3. revise VisitedList.4. use get_linklist_at_level to keep the code more tidy.5. add a simple example, it is tested under Linux.first, copy train-labels.idx1-ubyte, train-images.idx3-ubyte, t10k-labels.idx1-ubyte, t10k-images.idx3-ubyte to the same path of the executable file.then, run ./mnistit will look like this (we set k=1 here): efConstruction=48, M=16, vecdim=784Building index:Build time: 1.320725stesting...ef=8, recall=0.942000, time=83.650msef=9, recall=0.944700, time=89.679msef=10, recall=0.947100, time=93.547msef=11, recall=0.949000, time=116.838msef=12, recall=0.950700, time=104.603msef=13, recall=0.951600, time=111.117msef=14, recall=0.952500, time=115.467msef=15, recall=0.953300, time=120.319msef=16, recall=0.954000, time=124.361msef=17, recall=0.955600, time=148.542msef=18, recall=0.956300, time=154.498msef=19, recall=0.957100, time=138.762msef=20, recall=0.958000, time=142.354msef=21, recall=0.958500, time=150.261msef=22, recall=0.959100, time=150.379msef=23, recall=0.959900, time=154.988msef=24, recall=0.960700, time=159.845msef=25, recall=0.961100, time=177.775msef=26, recall=0.961700, time=167.215msef=27, recall=0.962100, time=171.925msef=28, recall=0.962700, time=179.467msef=29, recall=0.962800, time=179.995msef=30, recall=0.963100, time=184.014msef=31, recall=0.963800, time=188.360msActual memory usage: 115Mb I use an old desktop, if you have a new machine, it will be faster.15.03.2021, 11:20, "Yury Malkov" ***@***.***>:
Thank you! Looks good to me!
Will merge it after we figure our why CI flag is not showing in github web interface.

**Pull request closed without merging**